### PR TITLE
fix: cni-server default encap ip use right interface ip

### PR DESCRIPTION
**Pull request by Icedroid**
When a k8s node use multiple card, the cni server started for an incorrect interface . If the config.iface is empty, will use the node ip to be tunnel interface encap ip. If the config.iface set, use the config.iface for tunnel interface.

### CODE REVIEW

File: pkg/daemon/config.go [hunk 1]
```diff
@@ -122,20 +122,44 @@ func ParseFlags() (*Configuration, error) {
...
+		encapIP string
+	)
+
 	if config.Iface == "" {
 		i, err := getDefaultGatewayIface()
 		if err != nil {
 			return err
 		} else {
-			config.Iface = i
+			iface, err = findInterface(i)
```
Comment (oilbeater) [inline pkg/daemon/config.go:136]: the `config.Iface` here should be the interface that owns the `POD_IP`, not the default gateway

**Review by oilbeater — CHANGES_REQUESTED** [review]

File: pkg/daemon/config.go (hunk 1 above)
Comment (Icedroid) [reply pkg/daemon/config.go:136]: The default gateway only for mtu, must use the interface that owns the POD_IP?

Comment (oilbeater) [reply pkg/daemon/config.go:136]: yes, the packet will be tunneled by the interface that owns the encapIP, so the mtu should also coordinate with the same interface

File: pkg/daemon/config.go [hunk 2]
```diff
@@ -238,6 +255,31 @@ func getDefaultGatewayIface() (string, error) {
 	return "", errors.New("unable to find default route")
 }
 
+func getIfaceOwnPodIP(podIP string) (string, error) {
+	routes, err := netlink.RouteList(nil, syscall.AF_INET)
```
Comment (oilbeater) [inline pkg/daemon/config.go:259]: you can use `netlink.LinkList` and `netlink.AddrList` to retrieve the ip address of each interface. The route way might be incorrect if no `src` field is set

**Review by oilbeater — CHANGES_REQUESTED** [review]

**Review by Icedroid — COMMENTED** [review]
get interface owns POD_IP use netlink.LinkList

File: pkg/daemon/config.go [hunk 3]
```diff
@@ -238,6 +255,27 @@ func getDefaultGatewayIface() (string, error) {
...
+			return "", fmt.Errorf("failed to get a list of IP addresses %v", err)
+		}
+		for _, addr := range addrs {
+			if addr.IPNet.Contains(net.ParseIP(podIP)) && addr.IP.String() == podIP {
+				return link.Attrs().Name, nil
+			}
+		}
+	}
+
+	return "", errors.New("unable to find podIP interface")
```
Comment (oilbeater) [inline pkg/daemon/config.go:276]: it can return netlink directly here, no need to use `findInterface` later to find by name

**Review by oilbeater — CHANGES_REQUESTED** [review]

File: pkg/daemon/config.go (hunk 2 above)
Comment (Icedroid) [reply pkg/daemon/config.go]: Good idea.

**Comment by oilbeater** [comment]
@Icedroid  need to squash all commit into one commint, you can check https://www.internalpointers.com/post/squash-commits-into-one-git and use `git push -f` to force update the pr branch

**Review by oilbeater — APPROVED** [review]

**Comment by oilbeater** [comment]
Thanks! @Icedroid

**Pull request merged**